### PR TITLE
tools/cephfs: add pg_files command

### PR DESCRIPTION
**Pull request by jcsp**
tools/cephfs: add pg_files command
    
    Sometimes users know that particular data pool PGs
    have been damaged, and they would like to scan
    their files to work out which ones might have
    been affected.
    
    Fixes: http://tracker.ceph.com/issues/17249
    Signed-off-by: John Spray <john.spray@redhat.com>

### CODE REVIEW

**Comment by jcsp** [comment]
Checked this doesn't currently break anything in fs suite but it needs its own test too

File: src/tools/cephfs/PgFiles.cc [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+  int r = ceph_get_path_layout(cmount, path.c_str(), &l_stripe_unit,
+                               &l_stripe_count, &l_object_size,
+                               &l_pool_id);
+
+  struct file_layout_t layout;
+  layout.stripe_unit = l_stripe_unit;
+  layout.stripe_count = l_stripe_count;
+  layout.object_size = l_object_size;
+  layout.pool_id = l_pool_id;
+  if (r != 0) {
```
Comment (gregsfortytwo) [inline src/tools/cephfs/PgFiles.cc]: This delayed erro check isn't kinds weird?

File: src/tools/cephfs/PgFiles.cc [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+  ceph_dir_result *dr = nullptr;
+  int r = ceph_opendir(cmount, "/", &dr);
+  if (r != 0) {
+    derr << "Failed to open path: " << cpp_strerror(r) << dendl;
+    return;
+  }
+
+  struct dirent de;
+  while((r = ceph_readdir_r(cmount, dr, &de)) != 0) {;
```
Comment (gregsfortytwo) [inline src/tools/cephfs/PgFiles.cc]: Gratuitous semi-colon? Or is there some preventive thing going on?

File: src/tools/cephfs/PgFiles.cc [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        pg_num_mask = pool->get_pg_num_mask();
+        pg_num = pool->get_pg_num();
+      }
+    });
+    if (r != 0) {
+      // Can happen if layout pointed to pool not in osdmap, for example
+      continue;
+    }
+
+    target.m_seed = ceph_stable_mod(target.ps(), pg_num, pg_num_mask);
```
Comment (gregsfortytwo) [inline src/tools/cephfs/PgFiles.cc]: Are there no OSDMap or otherwise more "public" functions exposing this?

**Comment by gregsfortytwo** [comment]
My main concern here is that it will get used in place of a proper fsck sequence. It doesn't reconstruct the back traces on head objects, for instance.

Not sure if we should just note that in the docs or if there's a more definite care we can put in place (automaticallly re-creating any lost heads we detect here?)

File: src/tools/cephfs/PgFiles.cc (hunk 1 above)
Comment (jcsp) [inline src/tools/cephfs/PgFiles.cc]: Indeed it is, fixed.

File: src/tools/cephfs/PgFiles.cc (hunk 2 above)
Comment (jcsp) [inline src/tools/cephfs/PgFiles.cc]: Fixed

File: src/tools/cephfs/PgFiles.cc (hunk 3 above)
Comment (jcsp) [inline src/tools/cephfs/PgFiles.cc]: Nope, in existing use in Objecter it's not OSDMap that does this part

**Comment by jcsp** [comment]
I've updated this, and added a 'danger' block to the docs telling people to remove files and replace them rather than overwriting in place.  Ultimately we should also be repairing backtraces from forward scrub probably.

(Tested by https://github.com/ceph/ceph-qa-suite/pull/1216)

**Comment by jcsp** [comment]
New test passing here: http://pulpito.ceph.com/jspray-2016-10-20_14:39:41-fs-wip-jcsp-testing-20161020-distro-basic-mira/487477/

**Comment by jcsp** [comment]
@gregsfortytwo this should be good to go (and https://github.com/ceph/ceph/pull/11578 was inspired by this case too in case that wasn't obvious :-))

**Pull request merged**